addons/stock/models/stock_picking.py

```diff
@@ -845,7 +845,7 @@ def button_validate(self):
             picking_type = picking.picking_type_id
             precision_digits = self.env['decimal.precision'].precision_get('Product Unit of Measure')
             no_quantities_done = all(float_is_zero(move_line.qty_done, precision_digits=precision_digits) for move_line in picking.move_line_ids.filtered(lambda m: m.state not in ('done', 'cancel')))
-            no_reserved_quantities = all(float_is_zero(move_line.product_qty, precision_rounding=move_line.product_uom_id.rounding) for move_line in picking.move_line_ids)
+            no_reserved_quantities = all(float_is_zero(move_line.product_qty, precision_rounding=move_line.product_uom_id.rounding) and move_line.state != 'done' for move_line in picking.move_line_ids)
```

Comment:
As the issue happens with batch picking, I think it's better to fix this in `stock_batch_picking.py`.
In the `action_done`, we filter pickings to avoid done and cancel pickings, but just after, we call `button_validate` on all batch's pickings instead of the filtered ones. Calling `button_validate` on filtered pickings should do it.